festive-ferrets/backend/tictactoe/tictactoe.py

```diff
@@ -0,0 +1,128 @@
+        :param x: x - coordinate of the move
+        :type x: int
+        :param y: y - coordinate of the move
+        :type y: int
+        :return: If there is a winner, the winner, else None
+        """
+        if not self.board[x][y]:
+            self.board[x][y] = self.human
+            if self.check_winner():
+                return self.check_winner()
+            self.best_move()
+            if self.check_winner():
+                return self.check_winner()
+
+    def preview(self) -> List[List[str]]:
+        """
+        Returns the current board state
+        :return: Returns the current board state
+        """
+        return self.board
+
+    def best_move(self) -> None:
+        """
+        Make the best predicted move by the AI
+        :rtype: None
+        """
+        best_score = -float('inf')
+        move = None
+        for i in range(3):
+            for j in range(3):
+                if self.board[i][j] == '':
+                    self.board[i][j] = self.ai
+                    score = self.minmax(0, False)
+                    self.board[i][j] = ''
+                    if score > best_score:
+                        best_score = score
+                        move = (i, j)
+        self.board[move[0]][move[1]] = self.ai
+
+    def minmax(self, d: int, is_max: bool) -> Union[int, float]:
+        """
+        Return a minmax score for the state tree
+        :type d: int
+        :type is_max: bool
+        :return: Score for the state tree
+        """
+        result = self.check_winner()
+        if result:
+            return self.score[result]
+
+        if is_max:
+            best_score = -float('inf')
+            player = self.ai
+        else:
+            best_score = float('inf')
+            player = self.human
+        for i in range(3):
+            for j in range(3):
+                if self.board[i][j] == '':
+                    self.board[i][j] = player
+                    score = self.minmax(d + 1, False)
+                    self.board[i][j] = ''
+                    best_score = max(score, best_score)
+        return best_score
+
+    @staticmethod
+    def not_empty_eq(a: str, b: str, c: str) -> bool:
+        """
+        Returns True if all the variables are non empty strings and are equal
+        :type a: str
+        :type b: str
+        :type c: str
+        :return: Returns True if all the variables are non empty strings and are equal
+        :rtype: bool
+        """
+        return a == b == c and a != ''
+
+    def check_winner(self) -> Union[bool, str]:
+        """
+        Returns the winner if there is a winner for the current state
+        :return: False if no winner, 'X' or 'O' if either is winner
+        """
+        for i in range(len(self.board)):
+            if self.not_empty_eq(self.board[i][0], self.board[i][1], self.board[i][2]):
+                return self.board[i][0]
+
+        for i in range(len(self.board)):
```

Comment:
The third argument in the following line seems to have a fixed `0` instead of `i`. I think this is a bug:

```py
if self.not_empty_eq(self.board[0][i], self.board[1][i], self.board[2][0]):
```
should be
```py
if self.not_empty_eq(self.board[0][i], self.board[1][i], self.board[2][i]):
```

Another way of writing this is:
```py
for column in zip(*self.board):
    if self.not_empty_eq(*column):
        return column[0]
```